lib/Config/Config.php

```diff
@@ -0,0 +1,333 @@
+		'ldapUserDisplayName2'           => '',
+		'ldapGroupDisplayName'           => 'cn',
+		'ldapTLS'                        => 0,
+		'ldapQuotaDefault'               => '',
+		'ldapQuotaAttribute'             => '',
+		'ldapEmailAttribute'             => '',
+		'ldapGroupMemberAssocAttr'       => 'uniqueMember',
+		'ldapCacheTTL'                   => 600,
+		'ldapNetworkTimeout'             => 2,
+		'ldapUuidUserAttribute'          => 'auto',
+		'ldapUuidGroupAttribute'         => 'auto',
+		'homeFolderNamingRule'           => '',
+		'turnOffCertCheck'               => 0,
+		'ldapConfigurationActive'        => 0,
+		'ldapAttributesForUserSearch'    => '',
+		'ldapAttributesForGroupSearch'   => '',
+		'ldapExpertUsernameAttr'         => '',
+		'ldapExpertUUIDUserAttr'         => '',
+		'ldapExpertUUIDGroupAttr'        => '',
+		'hasMemberOfFilterSupport'       => 0,
+		'useMemberOfToDetectMembership'  => 1,
+		'lastJpegPhotoLookup'            => 0,
+		'ldapNestedGroups'               => 0,
+		'ldapPagingSize'                 => 500,
+		'ldapExperiencedAdmin'           => 0,
+		'ldapDynamicGroupMemberURL'      => '',
+		'ldapIgnoreNamingRules'          => false
+	];
+
+	public function __construct(array $data) {
+		$this->parseData($data);
+	}
+
+	/**
+	 * @return string the configuration prefix
+	 */
+	public function getId() {
+		return $this->id;
+	}
+
+	public function setId($id) {
+		$this->id = $id;
+	}
+
+	public function getData() {
+		return $this->data;
+	}
+
+	/**
+	 * @param string $name
+	 *
+	 * @return mixed|null
+	 */
+	public function __get($name) {
+		if (isset($this->data[$name])) {
+			return $this->data[$name];
+		}
+		return null;
+	}
+
+	/**
+	 * @param string $name
+	 *
+	 * @param mixed $value
+	 */
+	public function __set($name, $value) {
+		$this->setConfiguration([$name => $value]);
+	}
+
+	public function jsonSerialize() {
+		$data = [];
+		foreach ($this->data as $key=>$value) {
+			switch ($key) {
+				case 'ldapAgentPassword':
+					$value = \base64_encode($value);
+					break;
+				//following options are not stored but detected, skip them
+				case 'ldapIgnoreNamingRules':
+				case 'hasPagedResultSupport':
+				case 'ldapUuidUserAttribute':
+				case 'ldapUuidGroupAttribute':
+					continue 2;
+			}
+			if ($value === null) {
+				$value = '';
+			}
+			$data[$key] = $value;
+		}
+		return $data;
+	}
+
+	/**
+	 * set LDAP configuration with values delivered by an array, not read
+	 * from configuration. It does not save the configuration! To do so, you
+	 * must call saveConfiguration afterwards.
+	 * @param array $config array that holds the config parameters in an associated
+	 * array
+	 * @param array &$applied optional; array where the set fields will be given to
+	 * @return false|null
+	 */
+	public function setConfiguration($config, &$applied = null) {
+		if (!\is_array($config)) {
+			return false;
+		}
+
+		foreach ($config as $inputKey => $val) {
+			if (\array_key_exists($inputKey, $this->data)) {
```

Comment:
since `isset` is usually prefered over `array_key_exists`, include a comment saying that this usage is intended because some keys might have "null" as value.

Reply:
added